Add source module to create ProtonBunchIntensity sequence

brownd1978 opened the pull request:
There is still one issue, that the new ProtonBunchIntensitySummary object is not being put into the SubRun as it should.  I don't know if the problem is related with that (new) class or with the art source code.  That functionality is not essential for this pull.

FNALbuild commented:
Hi @brownd1978,
You have proposed changes to files in these packages:
- Sources
- MCDataProducts

which require these tests: build.




<a href="https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild">FNALbuild is explained here.</a>

brownd1978 commented:
Here's what's produced by ProtonBunchIntensityAnalyzer, to be compared with slide10 of doc 33344
[normal_PBIS.pdf](https://github.com/Mu2e/Offline/files/4839793/normal_PBIS.pdf)

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/212/commits/c10a766589a6988565c236103c6a0b69730d2a47: build

FNALbuild commented:
:sunny: The tests passed at ref c10a766589a6988565c236103c6a0b69730d2a47.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/scons.log). Build time: 11 min 51 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 3 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/fixme_todo.log) |
| clang-tidy | :heavy_check_mark: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//150/console).

kutschke commented on Sources/test/PBISequence.fcl:
```diff
@@ -0,0 +1,28 @@
+
+Output : {
+  module_type : RootOutput
+  outputCommands:   [ "drop *_*_*_*", "keep mu2e::ProtonBunchIntensity_*_*_*" ]
```
Please extend the keep command so that you keep the Subrun summary object.  For example:

  outputCommands:   [ "drop *_*_*_*",
                      "keep mu2e::ProtonBunchIntensity_*_*_*",
                      "keep mu2e::ProtonBunchIntensitySummary_*_*_*" ]

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/212/commits/f6c7548de0a0fc84417f1921450540311b4f5a92: build

brownd1978 commented:
Good catches!  The extra empty event is because the ifstream still returns 'good' after reading the last entry, I think because there's a final EOL character the << doesn't read.  I just moved the test to after reading the proton count.  I also replaced the files in /pnfs.

FNALbuild commented:
:sunny: The tests passed at ref f6c7548de0a0fc84417f1921450540311b4f5a92.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/scons.log). Build time: 13 min 32 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 3 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/fixme_todo.log) |
| clang-tidy | :heavy_check_mark: | [0 errors 0 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//151/console).

kutschke commented:
With these fixes it looks good to me and I will merge.   I have few things that we can follow-up on later.

About your comment in the code: "reconstitutes" is the way you spell "produces" in a source module.  I searched for "reconstitutes" in the art wiki - there is nothing useful for people who don't already know.  I will mention it to Kyle.

Online help does not work:

  mu2e --print-description PBISequence

Does not print the parameters of the detail class.   I will ask Kyle if we need to change:
   
    typedef art::Source<mu2e::PBISequenceDetail> PBISequence;

to get a module that is aware of a Config struct defined in the detail class.

I also checked what happens if I read two files sequentially.   We start new counter when we encounter the second file.  I like that as a default behavior.  In the future we may wish to add some configurability here - but no need to do that unless and until we need it.
